src/typechecker.c

```diff
@@ -901,39 +902,41 @@ NODISCARD bool typecheck_expression(AST *ast, Node *expr) {
 
       // TODO: Is complete to incomplete allowed?
 
-      // ALLOWED
-      // FROM any type T TO type T
+      // FROM any type T TO type T is ALLOWED
       if (types_equal(t_to, t_from)) break;
-      // FROM any pointer type TO any pointer type
+
+      // FROM any incomplete type is DISALLOWED
+      if (type_is_incomplete(t_from) || type_is_incomplete(t_to))
+        ERR(expr->cast.value->source_location, "Can not cast from an incomplete type %T", t_from);
+      // TO any complete type is DISALLOWED
+      if (type_is_incomplete(t_from) || type_is_incomplete(t_to))
+        ERR(expr->cast.value->source_location, "Can not cast to an incomplete type %T", t_to);
```

Comment:
You’re checking the same condition twice here

Reply:
Fixed 49c9970609854d8df692afdca587833e0e204ed0